### PR TITLE
Add httpClientBuilder attribute to driver builder to allow customized http client

### PR DESCRIPTION
**Pull request by guyilin-amazon**
## Description

Add `httpClientBuilder` attribute to driver builder to allow customized http client

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 1]
```diff
@@ -168,4 +169,6 @@
      * @return This builder object.
      */
     QldbDriverBuilder readAhead(int readAhead, ExecutorService executorService);
+
+    QldbDriverBuilder httpClientBuilder(SdkHttpClient.Builder httpClientBuilder);
```
Comment (simonz-bq) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:173]: Nit: Why is this prefixed with `SdkHttpClient`?

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 2]
```diff
@@ -168,4 +169,6 @@
      * @return This builder object.
      */
     QldbDriverBuilder readAhead(int readAhead, ExecutorService executorService);
+
```
Comment (simonz-bq) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:172]: I think we should take this chance to improve the docstrings. For one, this new builder option does not have a docstring.

On the SessionClientBuilder we should mention that the http client is overwritten, and a customized http wants to be used it needs to be passed into the new method, as well as briefly mentioning that the amount of connections should be equal or greater than the number of max concurrent transactions.

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java (hunk 1 above)
Comment (guyilin-amazon) [reply src/main/java/software/amazon/qldb/QldbDriverBuilder.java:173]: It's using `SdkHttpClient.Builder` defined in https://github.com/aws/aws-sdk-java-v2/blob/58c7300cd3d32f5c91c8f9e8b4f774826153321b/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpClient.java#L62

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java (hunk 2 above)
Comment (guyilin-amazon) [reply src/main/java/software/amazon/qldb/QldbDriverBuilder.java:172]: Will update the docstrings later.

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 3]
```diff
@@ -15,6 +15,7 @@
 
 import com.amazon.ion.IonSystem;
 import java.util.concurrent.ExecutorService;
+import software.amazon.awssdk.http.SdkHttpClient;
 import software.amazon.awssdk.services.qldbsession.QldbSessionClientBuilder;
 
 /**
```
Comment (allimn) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:21]: Do we need to update the snippet? We can build with a customized http client with max connections configured, which is helpful to differentiate with pool limit.

File: src/test/software/amazon/qldb/QldbDriverImplTest.java [hunk 4]
```diff
@@ -123,6 +124,29 @@ public void testBuildWitNullIonSystem() {
...
+        assertThrows(NullPointerException.class,
+                     () -> QldbDriver.builder()
+                                     .sessionClientBuilder(mockBuilder)
+                                     .ledger(LEDGER)
+                                     .httpClientBuilder(null)
+                                     .build());
+    }
+
+    @Test
+    public void testBuildWithCustomizedHttpClientBuilder() {
```
Comment (alpian) [inline src/test/software/amazon/qldb/QldbDriverImplTest.java:138]: This test is what I call an "MD5-test". You're writing test code which explicitly states what the production code should do at a syntactic level. Rather verify that, once built, the actual HTTP client in use is the one we want, and not the one the other code path creates. 

That might be really hard to do. In that case, you do not need to write a test. The cost of the test outweighs the value of the test.

**Review by alpian — CHANGES_REQUESTED** [review]

File: src/test/software/amazon/qldb/QldbDriverImplTest.java (hunk 4 above)
Comment (guyilin-amazon) [reply src/test/software/amazon/qldb/QldbDriverImplTest.java]: Deleted test

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java (hunk 3 above)
Comment (guyilin-amazon) [reply src/main/java/software/amazon/qldb/QldbDriverBuilder.java:21]: Updated

File: src/test/software/amazon/qldb/QldbDriverImplTest.java [hunk 5]
```diff
@@ -123,6 +124,16 @@ public void testBuildWitNullIonSystem() {
...
     }
 
+    @Test
+    public void testBuildWithNullHttpClientBuilder() {
+        assertThrows(NullPointerException.class,
+                     () -> QldbDriver.builder()
+                                     .sessionClientBuilder(mockBuilder)
+                                     .ledger(LEDGER)
+                                     .httpClientBuilder(null)
+                                     .build());
```
Comment (alpian) [inline src/test/software/amazon/qldb/QldbDriverImplTest.java:134]: Not convinced of the utility of this test, keep it if you feel strongly about it.

**Review by alpian — DISMISSED** [review]

**Review by allimn — DISMISSED** [review]

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 6]
```diff
@@ -61,7 +63,9 @@
     /**
      * <p>Specify the low level QLDB session builder that should be used for accessing QLDB.</p>
      *
-     * <p>Note that the user agent suffix and retry count will be set on the specified session builder.</p>
+     * <p>Note that the user agent suffix and retry count will be set on the specified session builder. And the
+     * http client will be overwritten. To pass customized http client to the the {@link QldbDriver},
+     * use {@link #httpClientBuilder(SdkHttpClient.Builder)}</p>
```
Comment (simonz-bq) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:68]: Minor grammatical issues:


Note that the user agent suffix and retry count will be set on the specified session builder, and the http client will be overwritten. To pass a customized http client to the the {@link QldbDriver}, use {@link #httpClientBuilder(SdkHttpClient.Builder)}.

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 7]
```diff
@@ -168,4 +172,23 @@
...
      */
     QldbDriverBuilder readAhead(int readAhead, ExecutorService executorService);
+
+    /**
+     * <p>Specify the http client builder that should be used for making http requests.</p>
+     *
+     * <p>Note that if maximum connections is set in the http client, it should be equal or greater than
+     * {@link #maxConcurrentTransactions(int)} to avoid connection contentions and poor performance.</p>
+     *
+     * <p>If no parameter is specified then a default value would be used.</p>
```
Comment (simonz-bq) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:182]: Nit: will be used.

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java [hunk 8]
```diff
@@ -168,4 +172,23 @@
...
+     *
+     * <p>
+     *     The httpClientBuilder is optional.
+     * </p>
```
Comment (simonz-bq) [inline src/main/java/software/amazon/qldb/QldbDriverBuilder.java:186]: I think this part is unneeded as it just repeats the above

Comment (guyilin-amazon) [reply src/main/java/software/amazon/qldb/QldbDriverBuilder.java:186]: We have "The `field` is optional/mandatory" for every field https://github.com/awslabs/amazon-qldb-driver-java/blob/3e7f5a4f3498513fee9cd337b8a5f14c941bca9b/src/main/java/software/amazon/qldb/QldbDriverBuilder.java#L117

File: src/main/java/software/amazon/qldb/QldbDriverBuilder.java (hunk 7 above)
Comment (guyilin-amazon) [reply src/main/java/software/amazon/qldb/QldbDriverBuilder.java:182]: This is consistent with other field docstrings, like `ionSystem` https://github.com/awslabs/amazon-qldb-driver-java/blob/3e7f5a4f3498513fee9cd337b8a5f14c941bca9b/src/main/java/software/amazon/qldb/QldbDriverBuilder.java#L80

File: src/test/software/amazon/qldb/QldbDriverImplTest.java (hunk 5 above)
Comment (guyilin-amazon) [reply src/test/software/amazon/qldb/QldbDriverImplTest.java:134]: I don't feel strongly that we need this test, will remove it.

**Review by simonz-bq — APPROVED** [review]

**Pull request merged**